tests/mock_crust/churn.rs

```diff
@@ -51,126 +53,135 @@ fn random_churn<R: Rng>(rng: &mut R,
+            let group = nodes.iter().filter(is_recipient).map(TestNode::name).collect();
+            let _ = self.groups.insert(dst, group);
         }
+        self.messages.insert((data_id, msg_id, src, dst));
     }
-}
-
-fn target_group<'a>(nodes: &'a [TestNode], target: &Authority) -> Vec<&'a TestNode> {
-    let dst = target.to_destination();
-    nodes.iter().filter(|n| n.is_recipient(&dst)).collect_vec()
-}
-
 
-fn send_requests(group: Vec<&TestNode>,
-                 src: Authority,
-                 dst: Authority,
-                 data_id: DataIdentifier,
-                 message_id: MessageId) {
-    for node in &group {
-        let _ = node.inner.send_get_request(src, dst, data_id, message_id);
+    /// Verifies that all sent messages have been received by the appropriate nodes.
+    fn verify(mut self, nodes: &[TestNode]) {
+        // The minimum of the group lengths when sending and now. If a churn event happened, both
+        // cases are valid: that the message was received before or after that. The number of
+        // recipients thus only needs to reach a quorum for the smaller of the group sizes.
+        let group_sizes: HashMap<_, _> = self.groups
+            .iter_mut()
+            .map(|(dst, group)| {
+                let dst_dest = dst.to_destination();
+                let is_recipient = |n: &&TestNode| n.is_recipient(&dst_dest);
+                let new_group = nodes.iter().filter(is_recipient).map(TestNode::name).collect_vec();
+                let count = cmp::min(group.len(), new_group.len());
+                group.extend(new_group);
+                (*dst, count)
+            })
+            .collect();
+        let mut group_msgs_received = HashMap::new(); // The count of received group messages.
+        for node in nodes {
+            while let Ok(event) = node.event_rx.try_recv() {
+                if let Event::Request { request: Request::Get(data_id, msg_id), src, dst } = event {
+                    let key = (data_id, msg_id, src, dst);
+                    if dst.is_group() {
+                        assert!(self.groups
+                                    .get(&key.3)
+                                    .map_or(false, |entry| entry.contains(&node.name())),
+                                "Unexpected request for node {:?}: {:?}",
+                                node.name(),
+                                key);
+                        *group_msgs_received.entry(key).or_insert(0usize) += 1;
+                    } else {
+                        assert_eq!(node.name(), *dst.name());
+                        assert!(self.messages.remove(&key),
+                                "Unexpected request for node {:?}: {:?}",
+                                node.name(),
+                                key);
+                    }
+                }
+            }
+        }
+        for key in self.messages {
+            assert!(key.3.is_group(), "Failed to receive request {:?}", key);
```

Comment:
This assertion was non-obvious to me - I had to read the method a few times to notice that all messages directed to single nodes should have been removed in the loop above. A comment here could be nice, but that's not a blocker.